datalad/distribution/add.py

```diff
@@ -0,0 +1,109 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""High-level interface for adding dataset components
+
+"""
+
+__docformat__ = 'restructuredtext'
+
+import logging
+from datalad.interface.base import Interface
+from datalad.support.constraints import EnsureStr, EnsureNone
+from datalad.distribution.dataset import EnsureDataset, datasetmethod
+from datalad.support.param import Parameter
+from datalad.interface.common_opts import recursion_flag, recursion_limit, \
+    git_opts, annex_opts, annex_add_opts
+
+lgr = logging.getLogger('datalad.distribution.add')
+
+
+class Add(Interface):
+    """Add files/directories to an existing dataset.
+
+    Typically, files and directories to be added to a dataset would be placed
+    into a directory of a dataset, and subsequently this command can be used to
+    register this new content with the dataset. With recursion enabled,
+    files will be added to their respective subdatasets as well.
+
+    Alternatively, a source location can be given to indicate where to obtain
+    data from. If no `path` argument is provided in this case, the content will
+    be obtained from the source location and a default local name, derived from
+    the source location will be generated. Alternatively, an explicit `path`
+    can be given to override the default.
+
+    If more than one `path` argument and a source location are provided, the
+    `path` arguments will be sequentially used to complete the source URL/path
+    (be means of concatenation), and an attempt is made to obtain data from
+    those locations.
+
+
+    || REFLOW >>
+    By default all files are added to the dataset's :term:`annex`, i.e. only
+    their content identity and availability information is tracked with Git.
+    This results in lightweight datasets. If desired, the [PY: `to_git`
+    PY][CMD: --to-git CMD] flag can be used to tell datalad to inject files
+    directly into Git. While this is not recommended for binary data or large
+    files, it can be used for source code and meta-data to be able to benefit
+    from Git's track and merge capabilities. Files checked directly into Git
+    are always and unconditionally available immediately after installation of
+    a dataset.
+    << REFLOW ||
+
+    .. note::
+      Power-user info: This command uses :command:`git annex add`, :command:`git annex addurl`, or
+      :command:`git add` to incorporate new dataset content.
+    """
+
+    _params_ = dict(
+        dataset=Parameter(
+            args=("-d", "--dataset"),
```

Comment:
Makes sense to me, yes. I'd centralize the treatment of paths anyway. We
currently have `resolve_path` in dataset.py and the `normalize_path` in
gitrepo.py. These should go into one thing. The base should always be the
repo/dataset an object is operating on. So, with `-d` the given dataset
should be the base indeed. Will do that.

2016-07-06 17:31 GMT+02:00 Yaroslav Halchenko <notifications@github.com>:

> In datalad/distribution/add.py
> <https://github.com/datalad/datalad/pull/598#discussion_r69753819>:
>
> > +    PY][CMD: --to-git CMD] flag can be used to tell datalad to inject files
> > +    directly into Git. While this is not recommended for binary data or large
> > +    files, it can be used for source code and meta-data to be able to benefit
> > +    from Git's track and merge capabilities. Files checked directly into Git
> > +    are always and unconditionally available immediately after installation of
> > +    a dataset.
> > +    << REFLOW ||
> > +
> > +    .. note::
> > +      Power-user info: This command uses :command:`git annex add`, :command:`git annex addurl`, or
> > +      :command:`git add` to incorporate new dataset content.
> > +    """
> > +
> > +    _params_ = dict(
> > +        dataset=Parameter(
> > +            args=("-d", "--dataset"),
>
> We can agree to have a -d but while at it let's agree then on what effect
> it would have: e.g. in Python interface we 'normalize' the passed filenames
> so they are treated as from the top of the repository. Thus, we could make
> it behave similarly here -- if "-d" is provided, treat relative paths as
> relative to the top of the dataset... makes sense?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/datalad/datalad/pull/598/files/0f80fc53047219c45ce9df6f1825b0f36e21b418#r69753819>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKAw_d43ufQpzdNpUe0Vio4kPfPtg2sbks5qS8pPgaJpZM4JA8UV>
> .
>